text/0000-freeze-uninit.md

```diff
@@ -0,0 +1,217 @@
+
+Other than the validity property, `read_freeze` has the same preconditions as the `read_by_copy` operation (used to implement `core::ptr::read`), or a read from a place obtained by dereferencing `ptr`. 
+In particular, it must be aligned for `T`, nonnull, dereferenceable for `T` bytes, and the bytes not be covered by a mutable reference that conflicts with `ptr`. 
+The validity invariant of `T` is enforced after the bytes read by the intrinsic are frozen. 
+
+The intrinsic is not proposed to be `const`, however there is not believed to be a fundamental reason why it could not be defined `const` in the future if there are sufficient use cases for that. This RFC leaves this to a separate stabilization step regardless.
+
+Any initialized byte value chosen nondeterminstically may be chosen arbitrarily by the implementation (and, in particular, the implementation is permitted to deliberately pick a value that will violate the validity invariant of `T` or otherwise cause undefined behaviour, if such a choice is available). 
+
+
+## Library Functions
+
+### `core::ptr::read_freeze` 
+```rust
+// in module core::ptr
+pub unsafe fn read_freeze<T>(ptr: *const T) -> T{
+    core::intrinsics::read_freeze(ptr)
+}
+```
+
+The `core::ptr::read_freeze` function directly invokes the `read_freeze` intrinsic and returns the result. 
+
+The majority of the operation is detailed [above](#language-primitive), so the definition is not repeated here.
+
+The `read_freeze` functions on `*const T` and `*mut T` are method versions of the free function and may be defined in the same manner.
+
+### `MaybeUninit::<T>::freeze`
+
+```rust 
+// in module core::mem
+impl<T> MaybeUninit<T>{
+    pub fn freeze(self) -> Self{
+        unsafe{core::ptr::read_freeze(core::ptr::addr_of!(self))}
+    }
+}
+```
+
+`MaybeUninit::freeze` is a safe version of the `read_freeze` function, defined in terms of the language intrinsic. It is safe because the value is kept as `MaybeUninit<T>`, which is trivially valid. 
+To be used, Rust code would need to unsafely call `MaybeUninit::assume_init` and assert that the frozen bytes are valid for `T`.
+
+The remaining behaviour of the function is equivalent to the `read_freeze` intrinsic.
+
+### Example Alternative Implementation
+
+If an implementation decides to define `MaybeUninit::freeze` (or equivalent) as the compiler intrinsic, it is possible to implement `core::ptr::read_freeze` as follows:
+```rust 
+pub unsafe fn read_freeze(ptr: *const T) -> T {
+    ptr.cast::<MaybeUninit<T>>()
+        .read()
+        .freeze()
+        .assume_init()
+}
+```
+
+# Drawbacks
+[drawbacks]: #drawbacks
```

Comment:
Another drawback that should be discussed is that this will lead to valgrind errors in sound Rust code. Ideally there is some plan for how to enable valgrind to still detect UB due to use of uninit memory (it would be bad to lose this capability) while accepting code that correctly uses `freeze`.

Reply:
I would imagine Rust can have a `cfg` that makes `freeze` ops be [reported to valgrind](https://valgrind.org/docs/manual/manual-core-adv.html#manual-core-adv.clientreq), so valgrind can properly handle it once that feature is implemented in valgrind.

Reply:
I'm pretty sure valgrind already has false positives in sound Rust code because it doesn't understand `MaybeUninit`.

Reply:
Do you have an example of that?